src/web/mod.rs

```diff
@@ -507,6 +508,25 @@ fn redirect_base(req: &Request) -> String {
     }
 }
 
+/// Parse and URL into a iron::Url struct.
+/// When `queries` are given these are added to the URL,
+/// with empty `queries` the `?` will be omitted.
+pub(crate) fn parse_url_with_params<I, K, V>(url: &str, queries: I) -> Result<Url, Error>
+where
+    I: IntoIterator,
+    I::Item: Borrow<(K, V)>,
+    K: AsRef<str>,
+    V: AsRef<str>,
+{
+    let mut queries = queries.into_iter().peekable();
+    if queries.peek().is_some() {
```

Comment:
I *think* this will compile, and it seems easier to understand.
```suggestion
    I: ExactIterator,
    I::Item: Borrow<(K, V)>,
    K: AsRef<str>,
    V: AsRef<str>,
{
    if queries.len() != 0 {
```

Reply:
it depends on what we're trying to achieve :) 

This the `BTreeMap` is not an interator, this would mean changing: 
```rust
let url = ctry!(req, parse_url_with_params(&url_str, queries));
```
into
```rust
let url = ctry!(req, parse_url_with_params(&url_str, queries.iter()));
```

at the call sites. 

Personally I prefer the former, also the method signature matches `Url::parse_with_params` then.

Reply:
```

pub(crate) fn parse_url_with_params<I, K, V>(url: &str, queries: I) -> Result<Url, Error>
where
    I: IntoIterator,
    I::IntoIter: ExactSizeIterator,
    I::Item: Borrow<(K, V)>,
    K: AsRef<str>,
    V: AsRef<str>,
{...}
```

works :)

Reply:
thanks! yeah, this would have worked too, only limiting the possible `queries` a little.